frontend/src/Frontend/Crypto/Class.hs

```diff
@@ -21,6 +25,46 @@ data PactKey = PactKey
   , _pactKey_secret :: ByteString
   } deriving Show
 
+data MnemonicError =
+    MnemonicError_InvalidPhrase
+  | MnemonicError_NotEnoughWords
+  deriving (Show)
+
+class (Show (BIP39MnemonicError mnem)) => BIP39Mnemonic mnem where
+  type BIP39MnemonicError mnem
+  generateMnemonic :: MonadJSM m => m (Either Text mnem)
+  toMnemonic :: MonadJSM m => [Text] -> m (Either (BIP39MnemonicError mnem) mnem)
+  mnemonicToText :: mnem -> [Text]
+
+instance BIP39Mnemonic [Text] where
+  type BIP39MnemonicError [Text] = MnemonicError
+
+  generateMnemonic = fmap Right genMnemonic
+
+  -- Basic case where input type ([Text]) is the same as output type and we are just validating
+  toMnemonic sentence = case (length sentence == 12) of
+    False -> pure $ Left MnemonicError_NotEnoughWords
+    True -> do
+      ffor (liftJSM $ validateMnemonic $ T.unwords sentence) $ \case
+        False -> Left MnemonicError_InvalidPhrase
+        True -> Right sentence
+  mnemonicToText = id
+
+-- Derive a root key from mnemonic; mostly used for setup workflow
+class BIP39Root key where
+  type Sentence key
+  deriveRoot :: MonadJSM m => Password -> Sentence key -> m (Maybe key)
+
+instance BIP39Root PrivateKey where
+  type Sentence PrivateKey = [Text]
+  deriveRoot pwd sentence = liftJSM $
+    fmap hush $ generateRoot pwd $ T.unwords sentence
+    where hush = either (const Nothing) Just
```

Comment:
Is this needed?  Can probably grab the implementation from `Control.Errors`.

Reply:
Fixed